#### &#128337; Temps previst de dedicació: 15 minuts

# TEMA 1: Casos específics de funcions

## Unitat 2: Recursivitat

Segons el diccionari de l'Institut d'Estudis Catalans, _recursiu_ significa _que es pot repetir teòricament un nombre infinit de vegades mitjançant l’aplicació d’una mateixa regla_. En el nivell A2 hem tingut l'ocasió de treballar un concepte semblant que ens permet tornar a executar una part del codi: les estructures iteratives amb els bucles `for`, `while` i `do ... while`. Les tècniques d'iteració i de recursivitat, tanmateix, funcionen de manera diferent.

En un bloc iteratiu executem unes línies de codi de manera repetida un nombre de vegades. Mitjançant la recursivitat, en canvi, definim una funció i en el seu codi interior cridem aquesta mateixa funció. Sembla un concepte un pèl estrany, però el veurem clar si comparem el codi dels dos enfocaments.

In [1]:
// Function to calculate factorial for a number.
function iterativeFactorial(number) {
    let result = number;

    for (let i = number - 1; i > 0; i--) {
        result = result * i;
    }

    return result;
}

console.log(iterativeFactorial(7));
console.log(iterativeFactorial(8));
console.log(iterativeFactorial(9));

5040
40320
362880


En aquest codi hem definit una funció que rep un nombre sencer com a paràmetre i, mitjançant un procés iteratiu, calcula el seu factorial. Hem utilitzat un bucle `for`, però podríem haver-hi utilitzat qualsevol dels altres tipus de bucle que ja coneixem. Si analitzem l'execució d'aquest codi, veiem que calcula el factorial de 7 mitjançant el càlcul iteratiu `7 * 6 * 5 * 4 * 3 * 2 * 1`. És a dir, utilitza un bucle per repetir el càlcul tantes vegades com calgui.

Si en lloc de calcular el factorial de 7 li demanem que calculi el factorial de 8, durà a terme el càlcul iteratiu `8 * 7 * 6 * 5 * 4 * 3 * 2 * 1`.

I, si en lloc de calcular el factorial de 8 li demanem que calculi el factorial de 9, durà a terme el càlcul iteratiu `9 * 8 * 7 * 6 * 5 * 4 * 3 * 2 * 1`.

Podem observar un clar patró recursiu: el factorial de cada nombre no és altra cosa que la multiplicació d'aquest nombre pel factorial del nombre anterior. És a dir, el factorial de 8 és 8 multiplicat pel factorial de 7. I el factorial de 9 és 9 multiplicat pel factorial de 8. Traduïm aquest patró a codi JavaScript.

In [2]:
// Function to calculate factorial for a number.
function recursiveFactorial(number) {
    return number === 1 ? 1 : number * recursiveFactorial(number - 1);
}

console.log(recursiveFactorial(7));
console.log(recursiveFactorial(8));
console.log(recursiveFactorial(9));

5040
40320
362880


> <br>🎯&nbsp; **Vist anteriorment**
> 
> En el nivell A2 vam introduir l'operador condicional ternari `?:`. Recordeu que aquest operador avalua una condició que ha de ser certa o falsa (primer operant) i en funció del seu valor retorna el segon operant (quan el primer és vertader) o el tercer operant (quan el primer és fals). Per exemple, si volem que una variable `x` prengui el valor 1 si la variable `y` és major que 10 i el valor 0 en qualsevol altre cas, podem escriure `const x = y > 10 ? 1 : 0;`.
> <br><br>

Hem tornat a definir una funció que rep un nombre sencer com a paràmetre. Però en aquesta ocasió utilitzem una estructura recursiva que funciona exactament així:

- Si el nombre rebut és 1, retorna immediatament el valor 1.
- En qualsevol altre cas, retorna el resultat de multiplicar el nombre pel factorial d'aquest mateix nombre menys un.

Simplement hem traduït a codi el patró recursiu que havíem identificat. Per a això, necessitem que el codi de la funció inclogui una crida a la mateixa funció. Resulta gairebé màgic! En executar aquest codi veiem que el resultat és el mateix que el que vam obtenir amb el codi iteratiu.

La seqüència de crides que es duen a terme per a un `number` amb valor inicial 7 és la següent:

![Exemple de recursivitat](images/recursive_factorial.png)

- `recursiveFactorial(7)` rep com a paràmetre el número 7 i intenta calcular el producte de 7 pel resultat de cridar la mateixa funció `recursiveFactorial()` passant-li el paràmetre 6, per la qual cosa queda a l'espera de rebre aquest resultat.
- `recursiveFactorial(6)` rep com a paràmetre el número 6 i intenta calcular el producte de 6 pel resultat de cridar la mateixa funció passant-li el paràmetre 5, per la qual cosa queda a l'espera de rebre aquest resultat.
- `recursiveFactorial(5)` rep com a paràmetre el número 5 i intenta calcular el producte de 5 pel resultat de cridar la mateixa funció passant-li el paràmetre 4, per la qual cosa queda a l'espera de rebre aquest resultat.
- `recursiveFactorial(4)` rep com a paràmetre el número 4 i intenta calcular el producte de 4 pel resultat de cridar la mateixa funció passant-li el paràmetre 3, per la qual cosa queda a l'espera de rebre aquest resultat.
- `recursiveFactorial(3)` rep com a paràmetre el número 3 i intenta calcular el producte de 3 pel resultat de cridar la mateixa funció passant-li el paràmetre 2, per la qual cosa queda a l'espera de rebre aquest resultat.
- `recursiveFactorial(2)` rep com a paràmetre el número 2 i intenta calcular el producte de 2 pel resultat de cridar la mateixa funció passant-li el paràmetre 1, per la qual cosa queda a l'espera de rebre aquest resultat.
- `recursiveFactorial(1)` rep com a paràmetre el número 1 i retorna directament el número 1.
- `recursiveFactorial(2)` ja pot calcular el producte de 2 pel resultat anterior 1, i obté un 2 que retorna a `recursiveFactorial(3)`.
- `recursiveFactorial(3)` ja pot calcular el producte de 3 pel resultat anterior 2, i obté un 6 que retorna a `recursiveFactorial(4)`.
- `recursiveFactorial(4)` ja pot calcular el producte de 4 pel resultat anterior 6, i obté un 24 que retorna a `recursiveFactorial(5)`.
- `recursiveFactorial(5)` ja pot calcular el producte de 5 pel resultat anterior 24, i obté un 120 que retorna a `recursiveFactorial(6)`.
- `recursiveFactorial(6)` ja pot calcular el producte de 6 pel resultat anterior 120, i obté un 720 que retorna a `recursiveFactorial(7)`.
- `recursiveFactorial(7)`, finalment, ja pot calcular el producte de 7 pel resultat anterior 720, i obté el resultat final 5040.

Tot aquest flux de crides i respostes pot semblar un pèl complex, però per al processador del nostre ordinador no suposa cap problema i es duu a terme de manera gairebé immediata. Convé, això sí, assegurar-se que el nombre de crides recursives necessàries per obtenir el resultat final no sigui exageradament alt, ja que en aquest cas podríem arribar a desbordar la memòria de l'ordinador i no obtindríem el resultat desitjat. En l'exemple anterior, el nombre de crides recursives és de sis, la qual cosa no suposa cap problema.

> <br>🚨&nbsp; **Atenció**
> 
> Quan treballem amb bucles és fàcil crear, per error, un bucle infinit. Passa el mateix amb les funcions recursives. Hem d'assegurar que les crides recursives a la funció tinguin sempre un límit. En l'exemple anterior ho aconseguim en definir que, si se sol·licita el factorial d'1, no torni a cridar la mateixa funció, sinó que retorni directament el valor 1. Aquesta és la clàusula de sortida de la nostra funció. Sense ella, la funció es cridaria infinitament a ella mateixa i el nostre programa no funcionaria correctament.
> <br><br>

<br>

> <br>💫&nbsp; **Ampliació**
> 
> Què passa si la nostra funció `recursiveFactorial()` rep com a paràmetre un nombre negatiu? Llavors... la clàusula de sortida no arribarà mai a executar-se i la funció es cridarà a ella mateixa de manera infinita. En programació hem d'assegurar-nos sempre que el nostre codi sigui robust davant d'usos no previstos. Més endavant, en el nivell B2, treballarem la gestió d'errors i excepcions. Ara per ara, n'hi ha prou amb saber que, si la nostra funció `recursiveFactorial()` rep com a paràmetre un nombre menor que 1, el programa es bloquejarà i no funcionarà correctament.
> <br><br>